### ISSUE TITLE
Mixed implicit + Access Token-only authentication for SPA client

### ISSUE THREAD

**Issue by RPM1984** [IdentityServer/IdentityServer3#2328]
Hi guys, 

I'm playing around with the [JavaScript Implicit Client](https://github.com/IdentityServer/IdentityServer3.Samples/tree/master/source/Clients/JavaScriptImplicitClient) example + minimal WebHost example, which most closely fits with the setup i'm trying to achieve.

I've noticed the sample client has a *Access Token Only* button, which i would _assume_ would enable  the client to get an access token (no identity token) from IdSrv, to be used to call the API. I confirmed the scope being passed to the authorize endpoint is `read write`.

But when i click this button, it prompts the login page, so i'm a bit confused - unless i'm missing something? If i'm not requesting an identity token, why is the user being prompted to supply credentials.

In my real world scenario, i'll have a SPA app interacting with a Nancy API secured by IdSrv bearer authentication - but some API operations don't need a user/specific claim, but they do need an access token. Essentially i just want my client/SPA application to get the `read write` scope, without user interaction. Do i need to setup a secondary IdSrv flow for this approach?

Basically - when the user first comes to my site, i want the SPA app to get an access token to be able to call my public API operations (i want these operations locked down to only my clients, but they don't need a user identity), then if the user logs in (or tries to access a secure resource), then i can do the implicit flow to get the identity token (as well as a new access token i guess).

Can someone give me advise on how i can best achieve this, and point to any samples, or how i can mod this sample to suit my needs?

Love, love the product btw! Thanks guys :)

**Comment by brockallen** [comment]
> But when i click this button, it prompts the login page, so i'm a bit confused - unless i'm missing something? If i'm not requesting an identity token, why is the user being prompted to supply credentials.

Because for a user to give permissions to a client to use a resource on their behalf, the token service needs to know who the user is. So the API and thus the token service need to know who the user is. But when getting back just an access_token the client (strictly speaking) hasn't authenticated the user. IOW, the id_token is what provides authentication to the client.

**Comment by RPM1984** [comment]
@brockallen thanks for your response. That makes sense. So any advise on what i should be doing here then? Like i said, i want my SPA app to call my API operations without forcing the user to authenticate. So is my only option to not require these public API operations an access token? (in which case, anyone on the internet can hit my public operations).

**Comment by fubar-coder** [comment]
Maybe I have the same problem? I want to use both bearer and cookies authentication. The bearer authentication is used from a native client, while the cookies authentication should be used from the web app, which also contains the web API used from the client using bearer authentication.

IOW: I want a web app that contains both web pages and a web API and thus should allow both cookies and bearer authentication.

The problem is, that I can only seem to get either the cookies authentication (implicit flow) or the bearer authentication to work, but not both. Any idea what I've done wrong? Maybe you've got an example how I should/could do this?

**Comment by brockallen** [comment]
@RPM1984:

> Like i said, i want my SPA app to call my API operations without forcing the user to authenticate. 

Then why do you need a token at all? It's an anonymous invocation given the way you described it.

Authenticating the client software when the code is public (JS) is really hard -- anyone can read the code and get the token/key/whatever you use for the authentication.

**Comment by RPM1984** [comment]
@brockallen you're right, and that's what i originally suspected - i just thought there might be a technique i'm not aware of. But at the end of the day, the JS in the public zone therefore any key stored can be easily intercepted. The only other way i can think of is to write a thin server-side proxy, and use client credentials for server-to-server communicate between the backend proxy and the API. But not worth the overhead.

Anyway, thanks for your advice. This issue can be closed.

Thanks again! You and Dom rock! :guitar:

**Comment by brockallen** [comment]
Thx
